File: src/main/java/com/enonic/app/datakit/DumpManager.java
package com.enonic.app.datakit;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.util.Properties;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

import org.osgi.service.component.annotations.Component;

import com.enonic.xp.home.HomeDir;

@Component(immediate = true)
public class DumpManager
{
    private static final Pattern JSON_STRING_PATTERN = Pattern.compile( "\"([^\"]+)\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"" );

    private static final Pattern JSON_NUMBER_PATTERN = Pattern.compile( "\"([^\"]+)\"\\s*:\\s*(\\d+)" );

    public String list()
    {
        final Path dumpDir = getDumpDirectory();

        if ( !Files.isDirectory( dumpDir ) )
        {
            return "[]";
        }

        final List<String> entries = new ArrayList<>();

        try ( DirectoryStream<Path> stream = Files.newDirectoryStream( dumpDir ) )
        {
            for ( Path entry : stream )
            {
                final String json = buildEntryJson( entry );
                if ( json != null )
                {
                    entries.add( json );
                }
            }
        }
        catch ( IOException e )
        {
            throw new RuntimeException( "Failed to list dumps: " + e.getMessage(), e );
        }

        return "[" + String.join( ",", entries ) + "]";
    }

    public boolean delete( final String name )
    {
        if ( name == null || name.isEmpty() )
        {
            return false;
        }

        final Path dumpDir = getDumpDirectory();
        final Path target = dumpDir.resolve( name ).normalize();

        if ( !target.startsWith( dumpDir ) )
        {
            return false;
        }

        try
        {
            if ( Files.isDirectory( target ) )
            {
                deleteRecursively( target );
                return true;
            }

            final Path archive = dumpDir.resolve( name + ".zip" ).normalize();
            if ( Files.isRegularFile( archive ) )
            {
                Files.delete( archive );
                return true;
            }

            return false;
        }
        catch ( IOException e )
        {
            throw new RuntimeException( "Failed to delete dump '" + name + "': " + e.getMessage(), e );
        }
    }

    private Path getDumpDirectory()
    {
        return HomeDir.get().toFile().toPath().resolve( "data" ).resolve( "dump" );
    }

    private String buildEntryJson( final Path entry )
    {
        final String name = entry.getFileName().toString();

        if ( Files.isDirectory( entry ) )
        {
            return buildDirectoryEntry( entry, name );
        }

        if ( name.endsWith( ".zip" ) )
        {
            return buildArchiveEntry( entry, name.substring( 0, name.length() - 4 ) );
        }

        return null;
    }

    private String buildDirectoryEntry( final Path dir, final String name )
    {
        final Path dumpJson = dir.resolve( "dump.json" );
        final Path exportProperties = dir.resolve( "export.properties" );

        if ( Files.isRegularFile( dumpJson ) )
        {
            try
            {
                final String content = new String( Files.readAllBytes( dumpJson ), StandardCharsets.UTF_8 );
                final String xpVersion = extractJsonString( content, "xpVersion" );
                final String modelVersion = extractJsonValue( content, "modelVersion" );
                final String timestamp = extractJsonString( content, "timestamp" );
                final long size = getDirectorySize( dir );

                return jsonEntry( name, timestamp, "versioned", xpVersion, modelVersion, size );
            }
            catch ( IOException e )
            {
                return jsonEntry( name, getLastModified( dir ), "versioned", "", "", getDirectorySizeSafe( dir ) );
            }
        }

        if ( Files.isRegularFile( exportProperties ) )
        {
            try ( InputStream is = Files.newInputStream( exportProperties ) )
            {
                final Properties props = new Properties();
                props.load( is );
                final String xpVersion = props.getProperty( "xp.version", "" );

                return jsonEntry( name, getLastModified( dir ), "export", xpVersion, "", getDirectorySize( dir ) );
            }
            catch ( IOException e )
            {
                return jsonEntry( name, getLastModified( dir ), "export", "", "", getDirectorySizeSafe( dir ) );
            }
        }

        return null;
    }

    private String buildArchiveEntry( final Path archive, final String name )
    {
        try
        {
            final long size = Files.size( archive );
            String xpVersion = "";
            String modelVersion = "";
            String timestamp = getLastModified( archive );

            try ( ZipFile zip = new ZipFile( archive.toFile() ) )
            {
                final ZipEntry dumpJsonEntry = findDumpJson( zip );
                if ( dumpJsonEntry != null )
                {
                    try ( InputStream is = zip.getInputStream( dumpJsonEntry ) )
                    {
                        final String content = new String( is.readAllBytes(), StandardCharsets.UTF_8 );
                        xpVersion = extractJsonString( content, "xpVersion" );
                        modelVersion = extractJsonValue( content, "modelVersion" );
                        final String ts = extractJsonString( content, "timestamp" );
                        if ( !ts.isEmpty() )
                        {
                            timestamp = ts;
                        }
                    }
                }
            }

            return jsonEntry( name, timestamp, "archived", xpVersion, modelVersion, size );
        }
        catch ( IOException e )
        {
            return jsonEntry( name, getLastModified( archive ), "archived", "", "", -1 );
        }
    }

    private ZipEntry findDumpJson( final ZipFile zip )
    {
        final Enumeration<? extends ZipEntry> entries = zip.entries();
        while ( entries.hasMoreElements() )
        {
            final ZipEntry entry = entries.nextElement();
            if ( entry.getName().equals( "dump.json" ) || entry.getName().endsWith( "/dump.json" ) )
            {
                return entry;
            }
        }
        return null;
    }

    private long getDirectorySize( final Path dir )
        throws IOException
    {
        long size = 0;
        try ( DirectoryStream<Path> stream = Files.newDirectoryStream( dir ) )
        {
            for ( Path entry : stream )
            {
                if ( Files.isDirectory( entry ) )
                {
                    size += getDirectorySize( entry );
                }
                else if ( Files.isRegularFile( entry ) )
                {
                    size += Files.size( entry );
                }
            }
        }
        return size;
    }

    private long getDirectorySizeSafe( final Path dir )
    {
        try
        {
            return getDirectorySize( dir );
        }
        catch ( IOException e )
        {
            return -1;
        }
    }

    private void deleteRecursively( final Path path )
        throws IOException
    {
        if ( Files.isDirectory( path ) )
        {
            try ( DirectoryStream<Path> stream = Files.newDirectoryStream( path ) )
            {
                for ( Path entry : stream )
                {
                    deleteRecursively( entry );
                }
            }
        }
        Files.delete( path );
    }

    private String extractJsonString( final String json, final String key )
    {
        final Matcher matcher = JSON_STRING_PATTERN.matcher( json );
        while ( matcher.find() )
        {
            if ( matcher.group( 1 ).equals( key ) )
            {
                return matcher.group( 2 );
            }
        }
        return "";
    }

    private String extractJsonValue( final String json, final String key )
    {
        final String number = extractJsonNumber( json, key );
        if ( !number.isEmpty() )
        {
            return number;
        }
        return extractJsonString( json, key );
    }

    private String extractJsonNumber( final String json, final String key )
    {
        final Matcher matcher = JSON_NUMBER_PATTERN.matcher( json );
        while ( matcher.find() )
        {
            if ( matcher.group( 1 ).equals( key ) )
            {
                return matcher.group( 2 );
            }
        }
        return "";
    }

    private String getLastModified( final Path path )
    {
        try
        {
            return Files.getLastModifiedTime( path ).toInstant().toString();
        }
        catch ( IOException e )
        {
            return "";
        }
    }

    private String jsonEntry( final String name, final String timestamp, final String type, final String xpVersion,
                              final String modelVersion, final long size )
    {
        return "{" +
            "\"name\":" + jsonString( name ) + "," +
            "\"timestamp\":" + jsonString( timestamp ) + "," +
            "\"type\":" + jsonString( type ) + "," +
            "\"xpVersion\":" + jsonString( xpVersion ) + "," +
            "\"modelVersion\":" + jsonString( modelVersion ) + "," +
            "\"size\":" + size +
            "}";
    }

    private String jsonString( final String value )
    {
        if ( value == null || value.isEmpty() )
        {
            return "\"\"";
        }
        return "\"" + value.replace( "\\", "\\\\" ).replace( "\"", "\\\"" ) + "\"";
    }
}

File: src/main/kotlin/com/enonic/app/datakit/DumpManager.kt
package com.enonic.app.datakit

import com.enonic.xp.home.HomeDir
import com.google.common.io.ByteSource
import com.google.common.io.Files as GuavaFiles
import org.osgi.service.component.annotations.Component
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import kotlin.io.path.readText

private val DUMP_JSON_STRING_REGEX = """"([^"]+)"\s*:\s*"((?:[^"\\]|\\.)*)"""".toRegex()
private val DUMP_JSON_NUMBER_REGEX = """"([^"]+)"\s*:\s*(\d+)""".toRegex()

@Component(immediate = true)
class DumpManager {
    fun list(): String {
        val dumpDir = dumpDirectory()
        if (!Files.isDirectory(dumpDir)) return "[]"

        val entries = mutableListOf<String>()

        try {
            Files.newDirectoryStream(dumpDir).use { stream ->
                for (entry in stream) {
                    val json = buildEntryJson(entry) ?: continue
                    entries += json
                }
            }
        } catch (e: IOException) {
            throw RuntimeException("Failed to list dumps: ${e.message}", e)
        }

        return entries.joinToString(",", prefix = "[", postfix = "]")
    }

    fun delete(name: String?): Boolean {
        if (name.isNullOrEmpty()) return false

        val dumpDir = dumpDirectory()
        val target = dumpDir.resolveChildEntry(name) ?: return false

        try {
            if (Files.isDirectory(target)) {
                target.deleteRecursively()
                return true
            }

            val archive = dumpDir.resolve("$name.zip")
            if (Files.isRegularFile(archive)) {
                Files.delete(archive)
                return true
            }

            return false
        } catch (e: IOException) {
            throw RuntimeException("Failed to delete dump '$name': ${e.message}", e)
        }
    }

    fun download(name: String?): ByteSource? {
        if (name.isNullOrEmpty()) return null

        val dumpDir = dumpDirectory()
        val target = dumpDir.resolveChildEntry(name) ?: return null

        val archive = dumpDir.resolve("$name.zip")
        if (Files.isRegularFile(archive)) {
            return GuavaFiles.asByteSource(archive.toFile())
        }

        if (Files.isDirectory(target) && isValidDumpDirectory(target)) {
            return target.zipToByteSource()
        }

        return null
    }

    fun upload(name: String?, data: InputStream?): Boolean {
        if (name.isNullOrEmpty() || data == null) return false

        val dumpDir = dumpDirectory()
        val entry = dumpDir.resolveChildEntry(name) ?: return false
        val target = dumpDir.resolve("$name.zip")

        if (Files.exists(target) || Files.isDirectory(entry)) return false

        val tempFile = dumpDir.resolve(".$name.zip.tmp")

        try {
            if (!Files.isDirectory(dumpDir)) {
                Files.createDirectories(dumpDir)
            }
            try {
                saveStream(tempFile, data)
                Files.move(tempFile, target)
            } finally {
                Files.deleteIfExists(tempFile)
            }
            return true
        } catch (e: IOException) {
            throw RuntimeException("Failed to upload dump '$name': ${e.message}", e)
        }
    }

    private fun isValidDumpDirectory(dir: Path): Boolean {
        return Files.isRegularFile(dir.resolve("dump.json")) ||
            Files.isRegularFile(dir.resolve("export.properties"))
    }

    private fun buildEntryJson(entry: Path): String? {
        val name = entry.fileName.toString()

        return when {
            Files.isDirectory(entry) -> buildDirectoryEntry(entry, name)
            name.endsWith(".zip") -> buildArchiveEntry(entry, name.removeSuffix(".zip"))
            else -> null
        }
    }

    private fun buildDirectoryEntry(dir: Path, name: String): String? {
        val dumpJson = dir.resolve("dump.json")
        val exportProperties = dir.resolve("export.properties")

        if (Files.isRegularFile(dumpJson)) {
            return try {
                val content = dumpJson.readText()
                val xpVersion = extractJsonString(content, "xpVersion")
                val modelVersion = extractJsonValue(content, "modelVersion")
                val timestamp = extractJsonString(content, "timestamp")
                val size = dir.directorySize()

                jsonEntry(name, timestamp, "versioned", xpVersion, modelVersion, size)
            } catch (_: IOException) {
                jsonEntry(name, dir.lastModifiedOrEmpty(), "versioned", "", "", dir.directorySizeSafe())
            }
        }

        if (Files.isRegularFile(exportProperties)) {
            return try {
                val props = Properties()
                Files.newInputStream(exportProperties).use { input ->
                    props.load(input)
                }

                val xpVersion = props.getProperty("xp.version") ?: ""
                jsonEntry(name, dir.lastModifiedOrEmpty(), "export", xpVersion, "", dir.directorySize())
            } catch (_: IOException) {
                jsonEntry(name, dir.lastModifiedOrEmpty(), "export", "", "", dir.directorySizeSafe())
            }
        }

        return null
    }

    private fun buildArchiveEntry(archive: Path, name: String): String {
        return try {
            val size = Files.size(archive)
            var xpVersion = ""
            var modelVersion = ""
            var timestamp = archive.lastModifiedOrEmpty()

            ZipFile(archive.toFile()).use { zip ->
                val dumpJsonEntry = findDumpJson(zip) ?: return@use

                zip.getInputStream(dumpJsonEntry).reader(Charsets.UTF_8).use { reader ->
                    val content = reader.readText()
                    xpVersion = extractJsonString(content, "xpVersion")
                    modelVersion = extractJsonValue(content, "modelVersion")

                    val archiveTimestamp = extractJsonString(content, "timestamp")
                    if (archiveTimestamp.isNotEmpty()) {
                        timestamp = archiveTimestamp
                    }
                }
            }

            jsonEntry(name, timestamp, "archived", xpVersion, modelVersion, size)
        } catch (_: IOException) {
            jsonEntry(name, archive.lastModifiedOrEmpty(), "archived", "", "", -1)
        }
    }

    private fun findDumpJson(zip: ZipFile): ZipEntry? {
        val entries = zip.entries()

        while (entries.hasMoreElements()) {
            val entry = entries.nextElement()
            if (entry.name == "dump.json" || entry.name.endsWith("/dump.json")) {
                return entry
            }
        }

        return null
    }

    private fun extractJsonString(json: String, key: String): String {
        for (match in DUMP_JSON_STRING_REGEX.findAll(json)) {
            if (match.groupValues[1] == key) {
                return match.groupValues[2]
            }
        }

        return ""
    }

    private fun extractJsonValue(json: String, key: String): String {
        val number = extractJsonNumber(json, key)
        if (number.isNotEmpty()) return number
        return extractJsonString(json, key)
    }

    private fun extractJsonNumber(json: String, key: String): String {
        for (match in DUMP_JSON_NUMBER_REGEX.findAll(json)) {
            if (match.groupValues[1] == key) {
                return match.groupValues[2]
            }
        }

        return ""
    }

    private fun jsonEntry(
        name: String,
        timestamp: String,
        type: String,
        xpVersion: String,
        modelVersion: String,
        size: Long,
    ): String = buildString {
        append('{')
        append("\"name\":")
        append(jsonString(name))
        append(",\"timestamp\":")
        append(jsonString(timestamp))
        append(",\"type\":")
        append(jsonString(type))
        append(",\"xpVersion\":")
        append(jsonString(xpVersion))
        append(",\"modelVersion\":")
        append(jsonString(modelVersion))
        append(",\"size\":")
        append(size)
        append('}')
    }
}

private fun dumpDirectory(): Path =
    HomeDir.get().toFile().toPath().resolve("data").resolve("dump")
